### PR TITLE
Polars: add date and time components

### LINKED ISSUE #1755: Polars: year, month, timezone, offsets, etc
Polars: year, month, timezone, offsets, etc
**Description**
- **Constructor**: Transformation
- **Class**: Expr
- **DSL Variant:** Function: TemporalExpr
--- 
reference: [Polars APIs sheet](https://docs.google.com/spreadsheets/d/1w4H-qrSe9pkvNT2NhowfiIHNWXSseOyMX2nu5H6-RkQ/edit?gid=0#gid=0)

### PR DESCRIPTION
**Pull request by Shoeboxam**
Close #1755

### CODE REVIEW

File: python/src/opendp/extras/polars/__init__.py [hunk 1]
```diff
@@ -67,6 +67,12 @@ class DPExpr(object):
           - Fill missing values with provided value
         * - `lit <https://docs.pola.rs/api/python/stable/reference/expressions/api/polars.lit.html>`_
           - Return an expression representing a literal value 
+        * - `str.strptime <https://docs.pola.rs/api/python/stable/reference/expressions/api/polars.Expr.str.strptime.html>`_
+          - Parse a string into a date, time or datetime
+        * - `date components <https://docs.pola.rs/api/python/stable/reference/expressions/temporal.html>`_
+          - year, iso_year, quarter, month, week, weekday, day, ordinal_day
+        * - `time components <https://docs.pola.rs/api/python/stable/reference/expressions/temporal.html>`_
+          - hour, minute, second, millisecond, microsecond, nanosecond
```
Comment (mccalluc) [inline python/src/opendp/extras/polars/__init__.py:75]: Double backticks around these names?

File: rust/src/transformations/make_stable_expr/namespace_dt/expr_datetime_component/mod.rs [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+#[cfg(test)]
+mod test;
+
+/// Make a Transformation that returns an expression that extracts a datetime component.
+///
+/// # Arguments
+/// * `input_domain` - Expr domain
+/// * `input_metric` - The metric under which neighboring LazyFrames are compared
+/// * `expr` - The dt expression
```
Comment (mccalluc) [inline rust/src/transformations/make_stable_expr/namespace_dt/expr_datetime_component/mod.rs:23]:
```suggestion
/// * `expr` - The datetime expression
```

File: rust/src/transformations/make_stable_expr/namespace_dt/expr_datetime_component/mod.rs [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+    (ExprDomain, M): MetricSpace,
+    Expr: StableExpr<M, M>,
+{
+    let Expr::Function {
+        input: inputs,
+        function: FunctionExpr::TemporalExpr(temporal_function),
+        ..
+    } = expr
+    else {
+        return fallible!(MakeTransformation, "expected dt expression");
```
Comment (mccalluc) [inline rust/src/transformations/make_stable_expr/namespace_dt/expr_datetime_component/mod.rs:41]:
```suggestion
        return fallible!(MakeTransformation, "expected datetime expression");
```

File: rust/src/transformations/make_stable_expr/namespace_dt/expr_datetime_component/mod.rs [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+        Quarter => (DataType::Int8, Some(4)),
+        Month => (DataType::Int8, Some(12)),
+        Week => (DataType::Int8, Some(53)),
+        WeekDay => (DataType::Int8, Some(7)),
+        Day => (DataType::Int8, Some(31)),
+        OrdinalDay => (DataType::Int16, Some(366)),
+        Hour => (DataType::Int8, Some(24)),
+        Minute => (DataType::Int8, Some(60)),
+        Second => (DataType::Int8, Some(60)),
+        Millisecond => (DataType::Int32, None),
```
Comment (mccalluc) [inline rust/src/transformations/make_stable_expr/namespace_dt/expr_datetime_component/mod.rs:127]:
```suggestion
        Millisecond => (DataType::Int32, Some(1000)),
```
just continuing the pattern?

File: rust/src/transformations/make_stable_expr/namespace_dt/expr_datetime_component/test.rs [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+use super::*;
+
+#[test]
+fn test_make_expr_components() -> Fallible<()> {
+    let (lf_domain, lf) = example_dt_lf()?;
+    let lf_plan = parse_example_dt_lf(lf.clone()).with_columns([
+        col("time").dt().second(),
+        col("datetime").dt().ordinal_day(),
+        col("date").dt().quarter(),
```
Comment (mccalluc) [inline rust/src/transformations/make_stable_expr/namespace_dt/expr_datetime_component/test.rs:15]:
```suggestion
        col("time").dt().second().alias("second"),
        col("datetime").dt().ordinal_day().alias("day"),
        col("date").dt().quarter().alias("quarter"),
```
so the headers on the table make sense?

File: rust/src/transformations/make_stable_expr/namespace_dt/expr_datetime_component/test.rs [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+use crate::transformations::make_stable_expr::namespace_dt::test::{
+    example_dt_lf, parse_example_dt_lf,
+};
+use crate::transformations::make_stable_lazyframe;
+
+use super::*;
+
+#[test]
+fn test_make_expr_components() -> Fallible<()> {
+    let (lf_domain, lf) = example_dt_lf()?;
```
Comment (mccalluc) [inline rust/src/transformations/make_stable_expr/namespace_dt/expr_datetime_component/test.rs:11]: Slight preference to keep the example data in the same file, even at the cost of some redundancy. I want to be able to look at the test and see that it's doing the right thing, rather than making opaque assertions.

File: rust/src/transformations/make_stable_expr/namespace_dt/test.rs [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+            "2024-09-13 15:00:45".to_string(),
+            "2025-10-14 16:00:56".to_string(),
+            "2025-13-14 16:00:39".to_string()
+        ],
+        "date" => &[
+            "2023-08-12".to_string(),
+            "2024-09-13".to_string(),
+            "2025-10-14".to_string(),
+            "2025-13-14".to_string()
+        ]
```
Comment (mccalluc) [inline rust/src/transformations/make_stable_expr/namespace_dt/test.rs:35]: I don't think these four rows are making a test that's any stronger than just one line? I'm sort of confused about the times that are almost but not quite the same: Is that to exercise anything particular?

To avoid distracting the user, I'd suggest one really boring date: `2000-01-01 00:00:00`, and any others should try to hit particular edge cases or errors.

**Review by mccalluc — COMMENTED** [review]
Main comment is just to simplify the test fixture (unless there's something special about this set of times and dates that I'm not seeing.)

**Comment by Shoeboxam** [comment]
Rebased on top of the ExprDomain and SeriesDomain refactors, and comments addressed in the rebase, so new changes aren't in a separate commit.

File: rust/src/transformations/make_stable_expr/namespace_dt/expr_datetime_component/mod.rs [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+        IsoYear => (DataType::Int32, None),
+        Quarter => (DataType::Int8, Some(4)),
+        Month => (DataType::Int8, Some(12)),
+        Week => (DataType::Int8, Some(53)),
+        WeekDay => (DataType::Int8, Some(7)),
+        Day => (DataType::Int8, Some(31)),
+        OrdinalDay => (DataType::Int16, Some(366)),
+        Hour => (DataType::Int8, Some(24)),
+        Minute => (DataType::Int8, Some(60)),
+        Second => (DataType::Int8, Some(60)),
```
Comment (mccalluc) [inline rust/src/transformations/make_stable_expr/namespace_dt/expr_datetime_component/mod.rs:117]: Looking at the next file, I think quarter is 1-indexed, while hour is 0-indexed. Should the values here for time units be one less?

File: rust/src/transformations/make_stable_expr/namespace_dt/expr_datetime_component/test.rs [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+│ ---      ┆ ---                 ┆ ---        ┆ ---       ┆ ---          ┆ ---          │
+│ time     ┆ datetime[ms]        ┆ date       ┆ i8        ┆ i16          ┆ i8           │
+╞══════════╪═════════════════════╪════════════╪═══════════╪══════════════╪══════════════╡
+│ 00:01:02 ┆ 2000-01-02 03:04:05 ┆ 2000-01-02 ┆ 0         ┆ 2            ┆ 1            │
+└──────────┴─────────────────────┴────────────┴───────────┴──────────────┴──────────────┘"#;
+    assert_eq!(format!("{:?}", actual), expected);
+    Ok(())
+}
+
+fn test_temporal_op_schema<const L: usize>(df: DataFrame, ops: [TemporalFunction; L]) {
```
Comment (mccalluc) [inline rust/src/transformations/make_stable_expr/namespace_dt/expr_datetime_component/test.rs:53]: Slight preference to reserve `test_` for tests, and a helper like this might be 
```suggestion
fn assert_temporal_op_schema<const L: usize>(df: DataFrame, ops: [TemporalFunction; L]) {
```

Imagining something looking at the code for a little bit and at first wondering where the `#[test]` is.

**Review by mccalluc — APPROVED** [review]
There might be an innocuous off-by-one problem, and slight preference for a different name for the helper function, but nothing that needs to block if I'm confused.

File: rust/src/transformations/make_stable_expr/namespace_dt/expr_datetime_component/mod.rs (hunk 8 above)
Comment (Shoeboxam) [reply rust/src/transformations/make_stable_expr/namespace_dt/expr_datetime_component/mod.rs:117]: No, this tracks the number of unique values, not the max value.
